transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy

```diff
@@ -592,69 +647,110 @@ class MultidimensionalDataResourceService implements MultiDimensionalDataResourc
         // check if there are any observations for the concept
-        if (!exists(builder, conceptConstraint)) {
-            throw new InvalidQueryException("No observations found for concept path: ${conceptConstraint.path}")
+        if (!exists(builder, constraint)) {
+            throw new InvalidQueryException("No observations found for query")
         }
 
-        // check if the concept is truly numerical (all textValue are E and all numberValue have a value)
-        // all(A) and all(B) <=> not(any(not A) or any(not B))
-        def valueTypeNotNumericConstraint = new FieldConstraint(
-                operator: Operator.NOT_EQUALS,
+        at.collect {aggregateFieldType(it)}.unique().findAll().each {
+            def wrongValueTypeConstraint = new FieldConstraint(
+                    operator: Operator.NOT_EQUALS,
+                    field: valueTypeField,
+                    value: it,
+            )
+            if(exists(builder, new AndConstraint(args: [(Constraint) constraint, wrongValueTypeConstraint]))) {
+                throw new InvalidQueryException('One of the concepts/observations has the wrong type for this aggregation')
+            }
+        }
+
+        // check if the concept is truly numerical (all textValue are E and all numberValue have a value) or textual
+
+        def textTypeConstraint = new FieldConstraint(
+                operator: Operator.EQUALS,
+                field: valueTypeField,
+                value: ObservationFact.TYPE_TEXT,
+        )
+
+        def numberTypeConstraint = new FieldConstraint(
+                operator: Operator.EQUALS,
                 field: valueTypeField,
-                value: ObservationFact.TYPE_NUMBER
+                value: ObservationFact.TYPE_NUMBER,
         )
-        def textValueNotEConstraint = new FieldConstraint(
-                operator: Operator.NOT_EQUALS,
+
+        def textValueEConstraint = new FieldConstraint(
+                operator: Operator.EQUALS,
                 field: textValueField,
                 value: "E"
         )
-        def numberValueNullConstraint = new NullConstraint(
-                field: numberValueField
-        )
-        def notNumericalCombination = new Combination(
-                operator: Operator.OR,
-                args: [valueTypeNotNumericConstraint, textValueNotEConstraint, numberValueNullConstraint]
-        )
-        def conceptNotNumericalCombination = new Combination(
-                operator: Operator.AND,
-                args: [conceptConstraint, notNumericalCombination]
-        )
 
-        if (exists(builder, conceptNotNumericalCombination)) {
-            def message = 'One of the observationFacts had either an empty numerical value or a ' +
-                    'textValue with something else then \'E\''
-            throw new InvalidQueryException(message)
-        }
+        // Transmart currently does not allow null values in the ObservationFacts, but it could be extended to allow
```

Comment:
The database and hypercube code could allow null values in ObservationFact by basically removing the checks like this one. However the tabular results will insert nulls for absent observations, so the old core db does not distinguish between nulls and absent values. Therefore the HddTabularResultHypercubeAdapter treats nulls in a tabular result as absent observations. 

I asked some data people, and as far as they were aware null values are not used and also not supported by kettle. So it seemed best to me to also disallow null values in the hypercube, hence the exception in ObservationFact. I just added a comment to document this.

I changed the aggregate implementation to first just try to retrieve results and only do the checks if no result is found. I don't like the doing 3 potentially expensive queries when 1 would suffice. This changes the semantics a bit, since if there are both valid values and invalid values it will now give an answer based on the valid values while previously it would return an error.